### PR TITLE
Add submariner install option & walkthrough

### LINKED ISSUE #164: Gateway API/HttpRoute PoC that leverages submariner for a multi-cluster topology
image

Put together a proof of concept that leverages submariner (via the submariner add-on in ocm) to establish a flat network between nodes, as per the topology diagram above.
Place a Gateway (implementation is Istio) on 1 cluster, and deploy an application to 2 other clusters.
Figure out where the HttpRoute can/should be placed, and what resources are required to allow this topology i.e. ServiceExports, ServiceImports.

The goals of this PoC are:

- determine the feasibility of this topology architecture
- identify what's involved in setting it up
- what actions are the responsibility of the 'platform/gateway administrator' persona vs. what responsibilities the application developer has

### PR DESCRIPTION
**Pull request by david-martin**
Closes #164 

By default submariner is *not* installed when you run `local-setup`.
A new env var is added, `SUBMARINER`, when set will install the [submariner components](https://submariner.io/getting-started/architecture/) and establish a flat network between the clusters.

2 proof of concept walkthroughs have been added, based on the ocm walkthrough.

The first goes through the process of deploying a gateway and app to 2 clusters, then adding resiliency via submariner by exporting & importing the app Service between the 2 clusters.

The second uses 3 clusters (a hub and 2 workload clusters).
The Gateway is deployed to the hub and routes traffic to a ServiceImport.
The ServiceImport imports the app service from the 2 workload clusters.

Demo video of the 2nd walkthrough https://youtu.be/yWDkttHUXXs

### CODE REVIEW

**Comment by david-martin** [comment]
> The Gateway responds with a 503 like it can't route the the externalName Service in the hub

After coming back to this later, it's now giving a 200.
I suspect it may be negative dns caching internally in kubernets.
I'll try reproduce from scratch and see if the order of things needs to change to avoid this scenario (if that's what's happening)

**Comment by david-martin** [comment]
/unhold

Changed ordering of serviceimport creation.
Good to review both walkthroughs

**Comment by R-Lawton** [comment]
👀

File: hack/make/dependencies.make [hunk 1]
```diff
@@ -92,3 +94,7 @@ clusteradm: $(CLUSTERADM)
 $(CLUSTERADM):
 	test -s $(LOCALBIN)/clusteradm || curl -sL https://raw.githubusercontent.com/open-cluster-management-io/clusteradm/main/install.sh | INSTALL_DIR=bin bash -s -- $(CLUSTERADM_VERSION)
 
+.PHONY: subctl
```
Comment (R-Lawton) [inline hack/make/dependencies.make:97]: im getting a 404 error for subctl when its trying to download it

Comment (R-Lawton) [reply hack/make/dependencies.make:97]: compared the url against if you were to download manually and it looks the same i wonder is there a flag with curl that you need

Comment (david-martin) [reply hack/make/dependencies.make:97]: I see this issue too now but only in iterm.
I didn't see it in vscode terminal.
The value for uname -m is coming back as x86 vs arm
So it's a Mac issue with that installer script as the arm version of the cli doesn't exist.
We should create an issue upstream for it.

If you want a workaround on Mac, you can download the amd mac binary and extract it to the dir. The make target shouldn't try to redownload if it's already there

File: docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+```
+
+Verify a 200 is still returned when routed to either cluster
+
+```bash
+curl -Ik --resolve sub.replace.this:443:172.32.200.0 https://sub.replace.this
+curl -Ik --resolve sub.replace.this:443:172.32.201.0 https://sub.replace.this
+```
+
+## Know issues
```
Comment (R-Lawton) [inline docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md:365]: The template was updated to fix the spelling mistake it should be known

File: docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+```bash
+(export $(cat ./controller-config.env | xargs) && export $(cat ./aws-credentials.env | xargs) && make build-controller install run-controller)
+```
+
+### Create a gateway
+
+We know will create a multi-cluster gateway definition in the hub cluster. In `T1`, run the following: 
+
+**Important**: Make sure to replace `sub.replace.this` with a subdomain of your root domain.
```
Comment (R-Lawton) [inline docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md:113]: i think adding the ❗  symbol to important would make it easier to pick out like the notes at the top of the walkthrough

Comment (R-Lawton) [reply docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md:113]: Same in other walkthrough as well

File: docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md (hunk 2 above)
Comment (R-Lawton) [reply docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md:365]: Same in other walkthrough

File: docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    name: api
+    hostname: sub.replace.this
+    port: 443
+    protocol: HTTPS
+EOF
+```
+
+### Place the gateway
+
+To place the gateway, we need to add a placement label to gateway resource to instruct the gateway controller where we want this gateway instantiated. In `T1`, run:
```
Comment (R-Lawton) [inline docs/how-to/submariner-poc-2-gateways-resiliency-walkthrough.md:137]: Nit pick: Capitlisaing resource names in some places but not all

**Comment by R-Lawton** [comment]
Apart from the few comments walkthroughs work well. I think once we get the issue automating a lot of these steps in the local-setup a lot of the repetition can be removed

**Comment by R-Lawton** [comment]
Theres a images folder outside how-to where the paths of the images are for walkthroughs because its outside how-to the paths are broken wdyt about pulling it into how-to and add your image ?

**Comment by david-martin** [comment]
@R-Lawton I believe I've updated all instances of `gateway` to be capitalised
I've added the subctl install issue to the known issues section

File: hack/.kindUtils [hunk 5]
```diff
@@ -1,13 +1,35 @@
...
+kindGenExternalKubeconfig() {
+  # Generate a kubeconfig that uses the docker bridge network IP address of the cluster
+  # This is required for using the subctl cmd (for submariner)
+  local master_ip
+  mkdir -p ./tmp/kubeconfigs/external/
+  EXTERNAL_KUBECONFIG=./tmp/kubeconfigs/external/${cluster}.kubeconfig
+  cp ./tmp/kubeconfigs/${cluster}.kubeconfig ${EXTERNAL_KUBECONFIG}
+  master_ip=$(docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' "${cluster}-control-plane" | head -n 1)
+  yq -i ".clusters[0].cluster.server = \"https://${master_ip}:6443\"" "${EXTERNAL_KUBECONFIG}"
+  yq -i "(.. | select(. == \"kind-${cluster}\")) = \"${cluster}\"" "${EXTERNAL_KUBECONFIG}"
```
Comment (mikenairn) [inline hack/.kindUtils:12]: This has added a dependency on yq, if it's not installed, you get:

```
Set kubectl context to "kind-mgc-control-plane"
./hack/.kindUtils: line 11: yq: command not found
make: *** [Makefile:83: local-setup] Error 127
```

We should either add it to .bin, or call it out somewhere how to install it.

**Pull request merged**